### PR TITLE
VSR: CheckpointTrailer in grid blocks

### PR DESCRIPTION
**Pull request by sentientwaffle**
Instead of encoding/decoding the trailer into/from a `buffer`, encode/decode it directly to/from grid blocks.

Benefits:
- One less copy (of all trailer data) during every checkpoint.
- Once we have a grid pool, blocks can be acquired lazily, and when not used (which is most of the time, since the free set is _usually_ very small) they contribute to the grid cache.

To keep implementation simple (avoiding modifications to `ClientSessions.{encode,decode}`), one assumption this makes is that the `ClientSessions` trailer fits in a single grid block. (The ClientSessions trailer is fixed-size.) This was already true for all of our configurations.

### CODE REVIEW

File: src/ewah.zig [hunk 1]
```diff
@@ -61,98 +64,208 @@ pub fn ewah(comptime Word: type) type {
...
+
+        pub const Encoder = struct {
+            source_words: []const Word,
+            source_index: usize = 0,
+            /// The number of literals left over from the previous encode() call that still need to
+            /// be copied.
+            literal_word_count: usize = 0,
+
+            /// Returns the number of bytes written to `target_chunk` by this invocation.
+            pub fn encode(encoder_: *Encoder, target_chunk: []align(@alignOf(Word)) u8) usize {
```
Comment (sentientwaffle) [inline src/ewah.zig]: The reason to have `encode` take a single chunk -- rather than an array of chunks that can be filled in sequence by a single function call -- is that it will let the `CheckpointTrailer` acquire blocks from the grid pool as-needed rather than acquiring the worst-case number. (Leading to fewer evictions from the grid cache.)

File: src/ewah.zig [hunk 2]
```diff
@@ -61,98 +64,208 @@ pub fn ewah(comptime Word: type) type {
             return @as(Word, @bitCast(mark));
         }
 
+        pub const Decoder = struct {
+            target_words: []Word,
+            target_index: usize = 0,
+            source_literal_words: usize = 0,
```
Comment (matklad) [inline src/ewah.zig:73]: One thing that gives me pause here is that there isn't any safety checks to ensure that the entire
thing was decoded. Basically, there's no difference between a Decoder which is driven to completion,
and a Decoder that missed some chunks (or processed a chunk twice).

Perhaps it makes sense to add expected total length and a `.finish()` method just for assertions?

Ah, curious how that problem doesn't exist for Encoder! Yeah, I think we should add `Decoder.done`,
so that the caller can `assert(decoder.done())`.

File: src/ewah.zig [hunk 3]
```diff
@@ -61,98 +64,208 @@
...
+
+            /// Returns the number of *words* written to `target_words` by this invocation.
+            // TODO Refactor to return an error when `source_chunk` is invalid,
+            // so that we can test invalid encodings.
+            pub fn decode(
+                decoder_: *Decoder,
```
Comment (matklad) [inline src/ewah.zig]: Let's try harder to avoid `thing_` names, as that goes against our intended style. Here, I thnik we
could have

```zig
Ewah.decode_all
Ewah.decode_chunks
```

to simultaneously free up decoder name, and also emphasize the connection between two decoding
methods.

And maybe call _this_ method `fn decode_chunk`?

File: src/ewah.zig [hunk 4]
```diff
@@ -61,98 +64,208 @@
...
+                    );
+                    source_index += literal_word_count_chunk;
+                    target_index += literal_word_count_chunk;
+                    decoder_.source_literal_words =
+                        marker.literal_word_count - literal_word_count_chunk;
+                }
+                assert(source_index <= source_words.len);
+                assert(target_index <= target_words.len);
+
+                defer decoder_.target_index = target_index;
```
Comment (matklad) [inline src/ewah.zig]: Nifty! But maybe move this defer right to the `var target_index` declaration, so that the reader
immediately knows the final destination of that local variable?

File: src/vsr/checkpoint_trailer.zig [hunk 5]
```diff
@@ -134,18 +129,19 @@ pub fn CheckpointTrailerType(comptime Storage: type) type {
...
-                .buffer = buffer,
+                .blocks = blocks,
+                .block_bodies = block_bodies,
                 .block_addresses = block_addresses,
                 .block_checksums = block_checksums,
             };
         }
 
         pub fn deinit(trailer: *Self, allocator: mem.Allocator) void {
+            for (trailer.blocks) |block| allocator.free(block);
```
Comment (matklad) [inline src/vsr/checkpoint_trailer.zig]: Move this right before `allocator.free(trailer.blocks);` to keep reverse allocation order.

File: src/vsr/checkpoint_trailer.zig [hunk 6]
```diff
@@ -167,6 +163,24 @@
...
+            for (
+                trailer.block_bodies[0..chunk_count],
+                trailer.blocks[0..chunk_count],
+                0..,
+            ) |*block_body, block, block_index| {
+                const chunk_size = Chunk.size(.{
+                    .block_index = @intCast(block_index),
+                    .block_count = chunk_count,
+                    .trailer_size = trailer_size,
+                });
```
Comment (matklad) [inline src/vsr/checkpoint_trailer.zig:187]: I am wondering whether it makes sense to try to slice this by `chunk_size`, rather than by
statically known `chunk_size_max` --- the resulting chunks are only "true" during decoding. When
encoding, we conservatively size chunks, and the last one might not be full.

No strong feelings either way!

Another thing here is that, ideally, we'd compute `StreamChecksum` as we are filling in the chunks,
rather than as a separate path afterwards, so that we don't do two passes over the same data.
Though, given that we have one chunk anyway, not that it matters :)

File: src/vsr/grid.zig [hunk 7]
```diff
@@ -338,8 +338,21 @@ pub fn GridType(comptime Storage: type) type {
...
+                );
+
+                grid.free_set_checkpoint.size = 0;
+                for (free_set_chunks) |chunk| {
+                    grid.free_set_checkpoint.size +=
+                        @as(u32, @intCast(free_set_encoder.encode(chunk)));
+
+                    if (free_set_encoder.done()) break;
+                }
+                assert(free_set_encoder.done());
```
Comment (matklad) [inline src/vsr/grid.zig]: Alternatively, `else unreachage`, as we have a defer above anyway

File: src/vsr/checkpoint_trailer.zig (hunk 6 above)
Comment (sentientwaffle) [reply src/vsr/checkpoint_trailer.zig:187]: > I am wondering whether it makes sense to try to slice this by chunk_size, rather than by
> statically known chunk_size_max --- the resulting chunks are only "true" during decoding. When
> encoding, we conservatively size chunks, and the last one might not be full.

I don't understand what you mean by this. :thinking:

Comment (matklad) [reply src/vsr/checkpoint_trailer.zig:187]: Yeah, sorry, let me rephrase!

What I mean is that the code _looks_ like it slices the blocks into the chunks we actually are going to use. So, for every block, we return not just `block[@sizeOf(vsr.Header)..block_size_max]`, but cap that to the exact size computed by the `Chunk`.


But that's not actually true during encoding! Chunks are allocated conservatively, so encoder might not actually be able to use up all of the chunks. So, for the encoding step, `block_body.* = block[@sizeOf(vsr.Header)..][0..chunk_size];` is actually dead code --- the encoder disregards `chunk_size` info, and instead stops when it itself thinks it should to stop. 

This seems a bit confusing to me. Specifically, the fact that, after encoding, some memory in `chunks` might actually be uninitialized. 

The alternative here is to say that `chunks` doesn't even attempt to compute which fraction of each block is used, and returns a `[]const [chunk_size_max]u8`, leaving it up to the caller to track the actual usage per chunk.

Comment (sentientwaffle) [reply src/vsr/checkpoint_trailer.zig:187]: Aha, that makes sense, thank you!

I think I'll split `.chunks()` into two different functions, `encode_chunks` and `decode_chunks`. Then neither need a `size` passed in -- `encode_chunks` returns an array of all max-size chunks, and `decode_chunks` uses `trailer.size` to truncate the last chunk. And `decode_chunks` is only used for reading existing chunks, so the returned slice can be `const`.

Comment (matklad) [reply src/vsr/checkpoint_trailer.zig:187]: Ah, brilliant, I was only noticing something fishy here, but indeed it looks like what we have here is just two different function!

**Review by matklad — APPROVED** [review]
